uc-crux-llvm/src/UCCrux/LLVM/Overrides.hs

```diff
@@ -0,0 +1,184 @@
+-- for verification. They unsoundly under-approximate the environment. This
+-- helps symbolic execution reach more code.
+registerUnsoundOverrides ::
+  (ArchOk arch, IsSymInterface sym, HasLLVMAnn sym) =>
+  proxy arch ->
+  L.Module ->
+  ModuleTranslation arch ->
+  OverM Model sym LLVM ()
+registerUnsoundOverrides proxy llvmModule mtrans =
+  do
+    let llvmCtx = mtrans ^. transContext
+    let ?lc = llvmCtx ^. llvmTypeCtx
+    register_llvm_overrides llvmModule [] (unsoundOverrides proxy) llvmCtx
+
+------------------------------------------------------------------------
+
+-- ** Declarations
+
+unsoundOverrides ::
+  ( IsSymInterface sym,
+    HasLLVMAnn sym,
+    ArchOk arch,
+    ?lc :: TypeContext,
+    HasModel personality
+  ) =>
+  proxy arch ->
+  [OverrideTemplate (personality sym) sym arch rtp l a]
+unsoundOverrides arch =
+  -- NOTE(lb): The list here should match that in the "Config" module.
+  [ basic_llvm_override $
+      [llvmOvr| i32 @gethostname( i8* , size_t ) |]
+        (\memOps sym args -> Ctx.uncurryAssignment (callGetHostName arch sym memOps) args),
+    basic_llvm_override $
+      [llvmOvr| i8* @getenv( i8* ) |]
+        (\memOps sym args -> Ctx.uncurryAssignment (callGetEnv arch sym memOps) args)
+  ]
+
+------------------------------------------------------------------------
+
+-- ** Implementations
+
+callGetHostName ::
+  ( IsSymInterface sym,
+    HasLLVMAnn sym,
+    wptr ~ ArchWidth arch,
+    ArchOk arch,
+    ?lc :: TypeContext
+  ) =>
+  proxy arch ->
+  sym ->
+  GlobalVar Mem ->
+  RegEntry sym (LLVMPointerType wptr) ->
+  RegEntry sym (BVType wptr) ->
+  OverrideSim p sym ext r args ret (RegValue sym (BVType 32))
+callGetHostName _proxy sym mvar (regValue -> ptr) (regValue -> len) =
+  do
+    let hostname = "hostname"
+    let lenLt bv = liftIO (What4.bvSlt sym len =<< What4.bvLit sym ?ptrWidth bv)
+    lenNeg <- lenLt (BV.mkBV ?ptrWidth 0)
+    -- TODO(lb): Panic if ?ptrWidth is like 2 or something crazy
+    lenSmall <- lenLt (BV.mkBV ?ptrWidth (fromIntegral (BS.length hostname)))
+    Override.symbolicBranches
+      emptyRegMap
+      [ ( lenNeg,
+          Override.modifyGlobal mvar $ \_mem ->
+            unimplemented "callGetHostName" Unimplemented.GetHostNameNegativeSize,
+          Nothing
+        ),
+        ( lenSmall,
+          Override.modifyGlobal mvar $ \_mem ->
+            unimplemented "callGetHostName" Unimplemented.GetHostNameSmallSize,
+          Nothing
+        ),
+        -- TODO Check for name size
+        -- Otherwise, return a canned name
+        ( What4.truePred sym,
+          Override.modifyGlobal mvar $ \mem ->
+            liftIO $
+              do
+                let val = LLVMMem.LLVMValString hostname
+                let ty = LLVMMem.llvmValStorableType val
+                mem1 <- LLVMMem.storeRaw sym mem ptr ty noAlignment val
+                bv0' <- What4.bvLit sym (knownNat @32) (BV.mkBV (knownNat @32) 0)
```

Comment:
Is it possible to annotate the resulting value as unsound so that we can factor that into reporting?

Reply:
Is it possible to learn constraints from observing how the synthesized names are used that might influence what we generate?

Reply:
> Is it possible to annotate the resulting value as unsound so that we can factor that into reporting?

Yes, and we should do that!

> Is it possible to learn constraints from observing how the synthesized names are used that might influence what we generate?

Maybe! What would the goal be? Perhaps more branch coverage?

Reply:
Yes, the goal would be more branch coverage. I'm not sure that we need to do it right now, but it would be good to think about what that looks like to make sure we don't put ourselves in a bad spot for later.

I was imagining that you might annotate the returned string values and then inspect the path conditions produced by symbolic execution. If you see that there is a branch that depends on one of the annotated terms, we could inspect the condition to figure out what inputs would allow us to reach more paths.  I guess there are a few questions with that:
1. Does annotation prevent constant propagation (and e.g., simplification of branch conditions)
2. Do we want it to or not?
3. Can we improve branch coverage by returning strings of concrete length but symbolic contents? Would doing that be enough? If we did that, would we then be able to see the branching in the path condition to explore alternative values?

Reply:
> I'm not sure that we need to do it right now

I think there are a few other things that would lead to more coverage more easily in the short term, but I think this would be a great thing to circle back to. Here are some items that I would guess would have a bigger immediate impact:

- The ability to skip undefined functions including un-overriden libc functions and any library functions not included in the bitcode
- Additional underapproximate overrides
- Additional *sound* overrides (e.g. `strchr`)
- Support for generating/reasoning about global variables

All of those are just guesses, though.

> but it would be good to think about what that looks like to make sure we don't put ourselves in a bad spot for later.

Yeah, I think we'd be in a reasonably good place to support this: Registering overrides happens in each iteration of the "bugfinding loop" (the one depicted in the flowchart in the README), and at each new iteration the information learned from previous iterations (constraints) is available. So we could collect whatever data we like during simulation and feed it back to the overrides in the next iteration pretty easily.

> Does annotation prevent constant propagation (and e.g., simplification of branch conditions)

TBH I don't actually know the mechanism of annotations, so I can't answer this one.

> Do we want it to or not?

This is even further outside my understanding :)

> Can we improve branch coverage by returning strings of concrete length but symbolic contents? Would doing that be enough? If we did that, would we then be able to see the branching in the path condition to explore alternative values?

Seems likely, and that would be a pretty darn easy first step.

Reply:
> Is it possible to annotate the resulting value as unsound so that we can factor that into reporting?

Done!